### DISCUSSION TITLE
Add-PNPFile : The remote server returned an error: (401) Unauthorized.

### DISCUSSION THREAD

**Discussion by marcomendi30** [pnp/powershell#1912] [General]
Im trying to upload files to sharepoint online with powershell PnP commands and modern auth.

Can connect succesfully to my site (https://lci2.sharepoint.com/sites/MarcoTestSite)

but when try to upload files with Add-PNPFile the powershell console show this error: Add-PNPFile : The remote server returned an error: (401) Unauthorized.

can you help me to fix this?

**Comment by dc366** [comment]
I’m going to guess as you haven’t provided the exact command you are trying to use.

If you are writing to a default document library, the main folder is actually called `Shared Documents` (I can’t remember if the space is there or not). If you try to write to `Documents`, it will fail as that folder doesn’t exist.

**Reply by marcomendi30** [reply]
#Connect without prompt
$SiteURL="https://lci2.sharepoint.com" 
$clientid="b5f3a2b8-9521-4868-bf35-97f0ba576ae1" 
$clientsecret="PMLkjEagVdLt6HdkJcWqkBeaUiV/prxxp9GSuQozm9E=" 


#Connect to PNP Online 
Connect-PnPOnline -Url $SiteURL -ClientId $clientid -ClientSecret $clientsecret -WarningAction Ignore

with this can I connect and run get-pnpcontext successfuly

Add-PNPFile -Path "‪C:\M2\powershell-sharepoint.txt" -folder "testscript" (this is a folder)


but when I try to upload files with add-pnpfile receive  The remote server returned an error: (401) Unauthorized.

**Comment by marcomendi30** [comment]
Hi, I'm using this command to upload files to sharepoint with Pnp.

Add-PNPFile -Path "‪C:\M2\powershell-sharepoint.txt" -folder "testscript"

**Comment by bscheewe** [comment]
I'm having the same problem.  If I use a global admin account it works fine.  If I use an App Registration account with a secret or cert, I get the 401 error. 

The app account started with the Sites.Selected graph API permission, but I upped it to Sites.FullControl.All and still the same error. Deleted the app account and started over and still the same.  ☹️

**Reply by bscheewe** [reply]
I was setting the permissions on the Microsoft Graph area instead of the Sharepoint area.  Once I granted Sites.FullControl.All on the Sharepoint API the app is working now.  This permission however is far to generous and I want to lock it down to a specific site.  The Sites.Selected permission is supposed to accomplish this, however I'm not yet clear on how to assign the site to the app account.

![image](https://user-images.githubusercontent.com/57464630/223619955-b23173ef-b573-40df-a5df-2e8a99aae297.png)

**Reply by paulconeil** [reply]
It's been a while but I thought I'd follow up.  I've tried giving Sites.Selected "FullControl" permissions to the app.   It appears you have to do this by doing the Grant-PnPAzureADAppSitePermission with Write permissions to the site first, then you have to follow up with a Set-PnPAzureADappSitePermission to then change it to FullControl.     However, when I do this I'm still getting access denied/401.   I'm wondering if there's some amount of time that passes before the permission takes effect and will update if it starts working.

**Reply by paulconeil** [reply]
I've figured it out.  You need to do what I mentioned above (Grant Write, then Set FullControl afterward) but the site also has to have NoScript Set to False.    Through this process I also ran into several bugs related to registering or permissions grant commands using PnP.Powershell 2.1.1 and 2.2.0 so I wound up grabbing the 2.2.81 (nightly build) to get everything working.

$reg = Register-PnPAzureADApp -ApplicationName $appName `
                        -Tenant $tenant `
                        -CertificatePath $certPfx `
                        -CertificatePassword $certPass `
                        -GraphApplicationPermissions "User.Read.All" `
                        -SharePointApplicationPermissions "Sites.Selected" `
                        -Interactive




#Connect-PnPOnline $siteUrl -Interactive
Write-Host "Granting WRITE permissions to app on: $siteUrl"
$grant = Grant-PnPAzureADAppSitePermission -Permissions Write -Site $siteUrl -AppId $reg.'AzureAppId/ClientId' -DisplayName "SitesResourceSpecific_$dateUnique"
$reset = Set-PnPAzureADAppSitePermission -PermissionId $grant.Id -Permissions FullControl

Write-Host "Setting NoScript to FALSE (so we can do file uploads)"
Set-PnPSite -NoScriptSite $false

**Reply by damienvanrobaeys** [reply]
Hi @bscheewe did this work for you using add-pnpfile and a secret ?
